sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs

```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
 
                     var literal = (LiteralExpressionSyntax)c.Node;
 
-                    var numberWithoutSuffix = ClearNumberTypeSuffix(literal.Token.Text);
-
-                    var decimalParts = numberWithoutSuffix.Split('.');
-                    if (decimalParts.Length > 2)
-                    {
-                        return;
-                    }
-
-                    var hasIrregularPattern = decimalParts
-                        .SelectMany(part => part.Split('_'))
-                        .Select(x => x.Length)
-                        .Skip(1) // skip the first part (1_234 => 234)
-                        .Reverse().Skip(decimalParts.Length == 2 ? 1 : 0) // skip the last if there is a decimal (.234_5 => 234)
-                        .Distinct().Skip(1).Any(); // we expect to have only 1 size of pattern
-
-                    if (hasIrregularPattern)
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        internal static bool HasIrregularPattern(string numericToken)
         {
-            if (numberLiteral.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            var splitted = StripNumericPreAndSuffix(numericToken).Split(Dot);
+
+            // ignore multiple dots.
+            if (splitted.Length > 2)
+            {
+                return false;
+            }
+
+            var groups = splitted[0].Split(Underscore).Select(s => s.Length).ToArray();
+
+            var size = NotFound;
+
+            if (groups.Length > 1)
+            {
+                size = groups[1];
+
+                // first should not be bigger.
+                if (groups[0] > size)
+                {
+                    return true;
+                }
+
+                // all should have the same size, except the first, and the size should at least be 2.
+                if (size < 2 || groups.Skip(1).Any(g => g != size))
+                {
+                    return true;
+                }
+            }
+
+            if (splitted.Length == 1)
             {
-                return numberLiteral.Substring(0, numberLiteral.Length - 2);
+                return false;
             }
 
-            if (NumericTypeSuffix.Contains(char.ToUpperInvariant(numberLiteral[numberLiteral.Length - 1])))
+            var decimals = splitted[1].Split(Underscore).Select(s => s.Length).ToArray();
+
+            if (decimals.Length == 1)
             {
-                return numberLiteral.Substring(0, numberLiteral.Length - 1);
+                return false;
             }
 
-            return numberLiteral;
+            size = size == NotFound ? decimals[0] : size;
+
+            // the last should not be bigger
```

Comment:
by "last" do you mean first? you mean that `0xF_FF_FF` is ok even if the first group has a different number of digits, right?

Reply:
yep

Reply:
Changed this code to make the intent clear. Thanks for the indebt code-reviews. It has been a long time since my code has been checked that well!

Reply:
I'm glad you appreciate the comments and thanks for the patience :)